File: algorithms/Smallest Subsequence of Distinct Characters/Smallest Subsequence of Distinct Characters.cpp
class Solution {
public:
    string smallestSubsequence(string s) {
        
        string st="";
        unordered_map< char ,int> m;
        vector< bool> vis( 26,false);
        for( int i=0;i<s.size();i++) m[s[i]]++;
        
        stack< char> t;
        
        t.push(s[0]) , m[s[0]]--;
        st+=s[0];
        vis[s[0]-'a']=true;
        
        for( int i=1;i<s.size();i++){
            
            m[ s[i]]--;
            if(!vis[ s[i]-'a']){
                while( !t.empty() &&  m[t.top()] >0 && t.top() > s[i]){
                    st.pop_back();
                    vis[ t.top()-'a']=false;
                    t.pop();
                }
                t.push(s[i]);
                vis[s[i]-'a']=true;
                st=st+s[i];
            }
        }
        return st;
    }
};

File: scripts/algorithms/S/Smallest Subsequence of Distinct Characters/Smallest Subsequence of Distinct Characters.js
// Runtime: 50 ms (Top 94.83%) | Memory: 42.10 MB (Top 87.93%)

var smallestSubsequence = function(s) {
    // Initialize an empty stack to store the result
    let stack = [];

    // Iterate through each character in the input string
    for (let i = 0; i < s.length; i++) {
        let char = s[i];

        // If the character is already in the stack, skip it
        if (stack.indexOf(char) > -1) continue;

        // Check and pop elements from the stack to maintain lexicographical order
        while (
            stack.length > 0 &&
            stack[stack.length - 1] > char &&
            s.indexOf(stack[stack.length - 1], i) > i
        ) {
            stack.pop();
        }

        // Push the current character onto the stack
        stack.push(char);
    }

    // Join the characters in the stack to form the smallest subsequence
    return stack.join('');
};
